dronecaster.lua

```diff
@@ -57,6 +66,71 @@ done_init = false
 
 -- init & core
 --------------------------------------------------------------------------------
+
+-- midi
+function build_midi_device_list()
+  midi_devices = {}
+  for i = 1,#midi.vports do
+    local long_name = midi.vports[i].name
+    local short_name = string.len(long_name) > 15 and util.acronym(long_name) or long_name
+    table.insert(midi_devices,i..": "..short_name)
+  end
+end
+
+function midi_event(data)
+  -- global midi setting
+  if midi_enabled == 1 then
+    msg = midi.to_msg(data)
+    -- filter channel
+    if msg.ch == midi_channel then
+      if msg.type ~= "clock" then
+        tab.print(msg)
+      end
+      --note message
+      if msg.type == "note_on" then
+        params:set("note",msg.note)
+        -- amp control
+        print(midi_amp_control)
+        if midi_amp_control == 2 then
+          amp = math.min(msg.vel/100, 1)
+        end
+        if midi_amp_control == 3 then
+          amp = math.min(msg.key_pressure/100, 1)
+        end
+        if midi_amp_control ~= 0 then
+          params:set("amp",amp)
+        end
+      end
+      --cc message
+      if msg.type == "cc" then
+        if msg.cc == midi_amp_cc then
+          params:set("amp", (msg.val/127))
+        end
+        if msg.cc == midi_drone_cc then
+          minval = math.min(msg.val, #drones)
+          params:set("drone", minval)
+        end
+      end
+      -- program change message
+      if msg.type == program_change then
+        minval = math.min(msg.val, #drones)
```

Comment:
At first I thought this was really clever because it means if you know the exact program value you're sending out, you can make a mapping that should be resilient to changes in drones. But then I thought, wait a minute, what are we sorting them by? Because unless it's by date added, the next time @schollz makes another dronecaster album with 8 new drones, the order is gonna get all messed up 😅

Reply:
Yes, but beyond my scope. It only really matters if people are making complex midi sequences where they want to set a specific patch. Other synths with customizable patches have the same issue: you can not predict how the users will order them. Better to see it as a "relative" instead of an "absolute" control.

Reply:
Hmm, I'm not 100% sure I'm tracking. Isn't setting a specific patch the whole point here? 

I think this implementation is fine, though - I can't think of a better way to do it that's as nice and simple as this.

Reply:
So there's two ways to set the drone: via customizable CC (very generic) or Program Change (specific). I added Program Change because my Digitakt has a dedicated encoder for this: it's less work to set up than a CC.
To fix the issue I pointed requires looking into the drone ordering code and see if worth changing so that new drones get added at the end. It's probably alphabetic now. Not sure if worth it for something so minor.

Reply:
Yeah, I think PC is better too. I agree that it's not worth forcing the sort order, too. If someone's really picky about the stability of that mapping, they could just avoid updating.